crates/oxc_transformer/src/regexp/mod.rs

```diff
@@ -0,0 +1,237 @@
+//!
+//! Note: You will need to include a polyfill for the `RegExp` constructor in your code to have the correct runtime behavior.
+//!
+//! ### ES2015
+//!
+//! #### Sticky flag (`y`)
+//! - @babel/plugin-transform-sticky-regex: <https://babeljs.io/docs/en/babel-plugin-transform-sticky-regex>
+//!
+//! #### Unicode flag (`u`)
+//! - @babel/plugin-transform-unicode-regex: <https://babeljs.io/docs/en/babel-plugin-transform-unicode-regex>
+//!
+//! ### ES2018
+//!
+//! #### DotAll flag (`s`)
+//! - @babel/plugin-transform-dotall-regex: <https://babeljs.io/docs/en/babel-plugin-transform-dotall-regex>
+//! - Spec: ECMAScript 2018: <https://262.ecma-international.org/9.0/#sec-get-regexp.prototype.dotAll>
+//!
+//! #### Lookbehind assertions (`/(?<=x)/` and `/(?<!x)/`)
+//! - Implementation: Same as esbuild's handling
+//!
+//! #### Named capture groups (`(?<name>x)`)
+//! - @babel/plugin-transform-named-capturing-groups-regex: <https://babeljs.io/docs/en/babel-plugin-transform-named-capturing-groups-regex>
+//!
+//! #### Unicode property escapes (`\p{...}` and `\P{...}`)
+//! - @babel/plugin-transform-unicode-property-regex: <https://babeljs.io/docs/en/babel-plugin-proposal-unicode-property-regex>
+//!
+//! ### ES2022
+//!
+//! #### Match indices flag (`d`)
+//! - Implementation: Same as esbuild's handling
+//!
+//! ### ES2024
+//!
+//! #### Set notation + properties of strings (`v`)
+//! - @babel/plugin-transform-unicode-sets-regex: <https://babeljs.io/docs/en/babel-plugin-proposal-unicode-sets-regex>
+//! - TC39 Proposal: <https://github.com/tc39/proposal-regexp-set-notation>
+
+mod options;
+
+use std::borrow::Cow;
+
+pub use options::RegExpOptions;
+use oxc_ast::ast::*;
+use oxc_regular_expression::ast::{LookAroundAssertionKind, Pattern, Term};
+use oxc_semantic::ReferenceFlags;
+use oxc_traverse::Traverse;
+
+use crate::context::Ctx;
+
+pub struct RegExp<'a> {
+    ctx: Ctx<'a>,
+    options: RegExpOptions,
+}
+
+impl<'a> RegExp<'a> {
+    pub fn new(options: RegExpOptions, ctx: Ctx<'a>) -> Self {
+        Self { ctx, options }
+    }
+}
+
+impl<'a> Traverse<'a> for RegExp<'a> {
+    fn enter_expression(
+        &mut self,
+        expr: &mut Expression<'a>,
+        ctx: &mut oxc_traverse::TraverseCtx<'a>,
+    ) {
+        let Expression::RegExpLiteral(regexp) = expr else {
+            return;
+        };
+
+        let is_unsupported = self.has_unsupported_regular_expression_flags(regexp.regex.flags);
+
+        let mut pattern_source = Cow::Borrowed(match regexp.regex.pattern {
+            RegExpPattern::Raw(pattern)
+                if is_unsupported
+                    || self.has_unsupported_regex_syntax_raw(pattern, regexp.regex.flags) =>
+            {
+                pattern
+            }
+            RegExpPattern::Pattern(ref pattern)
+                if is_unsupported || self.has_unsupported_regular_expression_pattern(pattern) =>
+            {
+                regexp.regex.pattern.source_text(self.ctx.source_text)
```

Comment:
This is a bit problematic. It's fine for now, but what if in future we allow arbitrary user-defined transforms, and the user has altered the regex's AST to change the pattern? In that case we'd clobber their changes here. Ideally we need a `Pattern::to_string` method.

Maybe just add a comment about this for now.

Reply:
Oh actually, #5304 added `Display` trait impl to `Pattern`. So `regexp.regex.pattern.to_string()` should do it.

Reply:
> If we want to avoid that work in common case where the `Pattern` has _not_ been modified, we could add a `pattern_str: Option<Atom<'a>>` field to `RegExp`. If you change the `Pattern`, you'd set `pattern_str` to `None`.

Yes, but we may easily forget to set `pattern_str` to `None`

Reply:
I decided to change to `Pattern::to_string`

Reply:
@Dunqing Just beware to_string method isn't a lossless operation and using source text is preferable when the intent is to keep the output same as the input.

furthermore the `RegExpPattern:: source_text` now always uses to_string underneath if span is unset. This change will be added in #5431

Reply:
As an example one quirk of the to_string is the fact that we lose capitalization information of hex values

Reply:
> @Dunqing Just beware to_string method isn't a lossless operation and using source text is preferable when the intent is to keep the output same as the input.
> 
> furthermore the `RegExpPattern:: source_text` now always uses to_string underneath if span is unset. This change will be added in #5431

Ah, I just realized!

Reply:
Is it necessarily a problem if we lose cosmetic information? i.e. isn't it OK to change the regex as long as we maintain its original behavior? This is a transform, which means we change stuff!

Reply:
Well Ideally for the minifier and transformers, we should use custom printers to optimize for size and compatibility respectively, But I agree it doesn't matter in the case of transformers but other parts of the project still may need to rely on the original source to keep this information.